### PR TITLE
[2.7] bpo-31692: Add PYTHONSHOWALLOCCOUNT env var

### PR DESCRIPTION
**Pull request by vstinner**
bpo-31692, bpo-19527:

* Add a new PYTHONSHOWALLOCCOUNT environment variable, similar to
  the Python 3 "-X showalloccount" option
* When Python is compiled with COUNT_ALLOCS, the new
  PYTHONSHOWALLOCCOUNT environment variable now has to be set to dump
  allocation counts into stderr on shutdown. Moreover, allocations
  statistics are now dumped into stderr rather than stdout.
* Add @test.support.requires_type_collecting decorator: skip test if
  COUNT_ALLOCS is defined
* Fix tests for COUNT_ALLOCS: decorate some methods with
  @requires_type_collecting
* test_sys.test_objecttypes(): update object type when COUNT_ALLOCS
  is defined

https://bugs.python.org/issue31692

### CODE REVIEW

File: Doc/using/cmdline.rst [hunk 1]
```diff
@@ -663,3 +663,10 @@ if Python was configured with the ``--with-pydebug`` build option.
 
    If set, Python will print memory allocation statistics every time a new
    object arena is created, and on shutdown.
+
+.. envvar:: PYTHONSHOWALLOCCOUNT
+
+   If set and Python was compiled with ``COUNT_ALLOCS`` defined, Python will
```
Comment (serhiy-storchaka) [inline Doc/using/cmdline.rst]: It may be worth to document the changes in `Doc/c-api/typeobj.rst` too.

**Review by serhiy-storchaka — COMMENTED** [review]
Using the `requires_type_collecting` decorator LGTM. But I don't sure about adding PYTHONSHOWALLOCCOUNT. This looks like a new feature to me.

What if split this PR on two parts? The one fixes issues related to immortal types, the other is about additional output.

**Comment by vstinner** [comment]
Serhiy: "...  This looks like a new feature to me ..."

I prefer to discuss on the bpo: https://bugs.python.org/issue31692#msg303959

File: Doc/using/cmdline.rst (hunk 1 above)
Comment (vstinner) [reply Doc/using/cmdline.rst]: I prefer to keep the description of all environment variables at the same place. I added a reference to the env var in Doc/c-api/typeobj.rst.

**Review by serhiy-storchaka — APPROVED** [review]

**Review by benjaminp — APPROVED** [review]

**Pull request merged**